_posts/2018-09-07-redis-ruby-bloom-filter.md

```diff
@@ -0,0 +1,355 @@
+---
+layout: post
+title: "Implementing a custom Redis and in-memory bloom filter"
+date: 2018-09-05 12:00:00 -0800
+cover: /assets/images/bloom_filter.svg
+excerpt: We recently changed our bloom filter implementation by using a custom Redis and an in-memory bloom filter written in Ruby. We will go through few iterations at solving a real problem writing a custom bloom filter from scratch.
+options:
+  - full-bleed-cover
+authors:
+  - name: Dalibor Nasevic
+    url: https://dalibornasevic.com
+    photo: https://avatars.githubusercontent.com/dalibor
+---
+
+
+Recently, we had to do an optimization that involved replacing a bloom filter built in-memory and stored on Amazon S3 to a combination of a Redis bloom filter and an in-memory bloom filter. In this blog post we'll go through the reasoning about this change as well as the details of the bloom filter implementation we landed on. Let's first start with a very brief info about what is a bloom filter.
+
+
+### What is a bloom filter?
+
+[Bloom filter](https://en.wikipedia.org/wiki/Bloom_filter) is a space-efficient probabilistic data structure, designed to test whether an element is a member of a set. Because it is probabilistic in nature, it can only guess if an element is in a set with a certain precision or tell for sure if an element is not in a set. That is an important detail to design around as we'll see later. If you're curious about the math involved, check out this [blog post](https://www.igvita.com/2008/12/27/scalable-datasets-bloom-filters-in-ruby/) for few additional details.
+
+
+### What is the real problem we are solving?
+
+In our email delivery products ([GoDaddy Email Marketing](https://www.godaddy.com/online-marketing/email-marketing) and [Mad Mimi](https://madmimi.com)), per our plans, we have a limit of maximum unique contacts that our customers can send emails to in a billing cycle. An interesting abuse scenario happens when uploading a list of email addresses, sending a campaign to that list, and then deleting the list, importing another list with different email addresses and sending another campaign again. We call this scenario "deleting and replacing" and to prevent it we need to keep a history of contacts that have been emailed in a billing cycle.
+
+
+### The naive solution
+
+The naive solution would be to check against the deliveries history in a billing cycle. While that might work for smaller data sets, it's a huge performance problem when dealing with billions of contacts. That is where the opportunity for using bloom filter data structure emerges.
```

Comment:
I don't know what the deliveries history is, so I don't understand this section. Is there a way to make this more accessible to people who might not be familiar with that term?